File: src/main/java/Day1to9/Day1.java
package Day1to9;

import utility.ExecutionTime;
import utility.Utility;

public class Day1 {
    String test = "test";
    String real = "Day1to9\\day1";
    
//    List<Integer> input = Utility.readIntegerFile(real);
    int[] input = Utility.readIntegerFile(real).stream()
        .mapToInt(Integer::intValue)
        .toArray();
    
    public Day1(){
        ExecutionTime.start();
    
        System.out.println(part1(1));
        
        System.out.println(part1(3));
        
        ExecutionTime.stop();
    }
    
    public int part1(int offset) {
        int increases = 0;
        for (int i = offset; i < input.length; i++) {
            if (input[i-offset] < input[i]) {
                increases++;
            }
        }
        return increases;
    }
    
    
    public static void main(String[] args) {
        new Day1();
    }
}

File: src/main/kotlin/day1to9/Day1.kt
package day1to9

import utility.ExecutionTime
import utility.Utility

fun main() {
    ExecutionTime.start()
    val expenses = Utility.readIntegerFile("AoC2020.Day1to9\\puzzle1")
//        .sorted()
    expenses.sort()

    fun findTwoFactors(targetValue : Int) : IntArray {
        var smallIndex = 0
        var bigIndex = expenses.size -1
        for (i in expenses.indices) {
            val sum = expenses[smallIndex] + expenses[bigIndex]
            when {
                sum > targetValue -> bigIndex--
                sum < targetValue -> smallIndex++
                else -> return intArrayOf(expenses[smallIndex], expenses[bigIndex])
            }
        }
        throw NoSuchElementException()
    }

    fun findThreeFactors(targetValue: Int): IntArray {
        for (currentInt in expenses) {
            try {
                val twoFactors = findTwoFactors(targetValue - currentInt)
                if (twoFactors[0] != currentInt && twoFactors[1] != currentInt) {
                    return intArrayOf(currentInt, twoFactors[0], twoFactors[1])
                }
            } catch (e: NoSuchElementException) {
            }
        }
        throw NoSuchElementException()
    }

    val (twoFirst, twoSecond) = findTwoFactors(2020)
    println("Found $twoFirst and $twoSecond. Final product is " + twoFirst*twoSecond)
    val (threeFirst, threeSecond, threeThird) = findThreeFactors(2020)
    println("Found $threeFirst, $threeSecond and $threeThird. Final product is " + threeFirst*threeSecond*threeThird)

    ExecutionTime.stop()
}